### PR TITLE
implement APS layer crypto

### PR DESCRIPTION
**Pull request by AdamLaurie**
Note this has been created/tested using scapy-com with patches as per my PR https://bitbucket.org/secdev/scapy-com/pull-requests/7/9003c5f9bdf2/diff which fix fragmentation and byte order issues (I know this should move to the main scapy repo but that seems to be broken for zigbee in too many other ways for me to get involved with right now! :)

### CODE REVIEW

**Comment by rmspeers** [comment]
I agree I'd love to get this using the "new" Github repo for scapy, instead of the scapy-com old repos. I've posted in https://github.com/secdev/scapy/pull/1179#event-1554785802 just now to try to see if we can get that stuff merged down.

**Comment by gpotter2** [comment]
Hi !

I’m one of scapy’s maintainers, and currently working into 802.15.4 re-integration. We already have a buggy implementation of 802.15.4, working properly with 6LoWPAN, and the ZigBee one should be fixed after https://github.com/secdev/scapy/pull/1456

It should get merged quite soon, feel free to report any issue you have with it

**Comment by AdamLaurie** [comment]
I'm happy to test against mainstream scapy once it's settled down... Is there any reason not to merge this into killerbee in the meantime?

**Comment by AdamLaurie** [comment]
Hmmm... I've just noticed that some packets in the sample pcap file fail to match when tested with the test script. This appears to be due to incorrect calculation of MIC as the CRYPTO is fine. Test with:

./scripts/zbtestcrypt -k 26546b723b396a727b5d5271517d392f -r sample/control4-sample.pcap -v 3

[...]
224 of 407 packets tested
  NWK crypto passed: 188
  NWK crypto failed: 36
    APS crypto passed: 0
    APS crypto failed: 0

I will investigate what needs to be done to correct the zigbeeData field and revert (although it'll be a couple of weeks as I'm working on something else now!) so feel free to take a look in the meantime...

**Comment by AdamLaurie** [comment]
After a little poking around I think it's nothing to do with the new crypto layers but failure to detect a  ZigbeeAppDataPayload / ZigbeeClusterLibrary layer...

Here is a packet that works OK:

Packet: 3
   Dot15d4FCS / 802.15.4 Data ( 0x00:0xb7e4 -> 0x3359:0x18c0 ) / ZigbeeNWK / ZigbeeSecurityHeader
   <Dot15d4FCS  fcf_reserved_1=0 fcf_panidcompress=True fcf_ackreq=True fcf_pending=False fcf_security=False fcf_frametype=Data fcf_srcaddrmode=Short fcf_framever=0 fcf_destaddrmode=Short fcf_reserved_2=0 seqnum=128 |<Dot15d4Data  dest_panid=0x3359 dest_addr=0x18c0 src_addr=0xb7e4 |<ZigbeeNWK  discover_route=0 proto_version=2 frametype=data flags=security+extended_dst+extended_src destination=0x00 source=0xb7e4 radius=10 seqnum=234 ext_dst=ff:f0:00:01:f0:22 ext_src=ff:f0:00:04:15:b1 |<ZigbeeSecurityHeader  reserved1= extended_nonce=1 key_type=network_key nwk_seclevel=None fc=0x730c source=ff:f0:00:04:15:b1 key_seqnum=0 data='[\x9d6\xfc{\x10\t-\xffu,\xe8y\xbb\xcai\x9dR\xc5\xdd\x90\x8b\xd7\x87\xba\xb4/\\\x02:\xd4\xd8F' |>>>>
   decrypting NWK...
Decrypt Details:
        Key:            26546b723b396a727b5d5271517d392f
        Nonce:          1a5b410000ff0f000c7300002d
        ZigbeeData:     081a0000e4b70aea22021f0000ff0f001a5b410000ff0f002d0c7300001a5b410000ff0f0000
        Decrypted Data: 40c501005cc2c52c3074363437302073612063342e7a722e6d6f740d0a
        Encrypted Data: 5b9d36fc7b10092dff752ce879bbca699d52c5dd908bd787bab42f5c02
        Mic:            3ad4d846
      ZigbeeAppDataPayload / ZigbeeClusterLibrary / Raw
      <ZigbeeAppDataPayload  frame_control=ack_req delivery_mode=unicast aps_frametype=data dst_endpoint=197 cluster=power_configuration profile=49756 src_endpoint=197 counter=44 |<ZigbeeClusterLibrary  reserved=1 disable_default_response=1 direction=0 manufacturer_specific=0 zcl_frametype=0 transaction_sequence=116 command_identifier=54 |<Raw  load='470 sa c4.zr.mot\r\n' |>>>
      40c501005cc2c52c3074363437302073612063342e7a722e6d6f740d0a
   encrypting NWK...
Encrypt Details:
        Key:            26546b723b396a727b5d5271517d392f
        Nonce:          1a5b410000ff0f000c7300002d
        ZigbeeData:     081a0000e4b70aea22021f0000ff0f001a5b410000ff0f002d0c7300001a5b410000ff0f0000
        Decrypted Data: 40c501005cc2c52c3074363437302073612063342e7a722e6d6f740d0a
        Encrypted Data: 5b9d36fc7b10092dff752ce879bbca699d52c5dd908bd787bab42f5c02
        Mic:            3ad4d846
   <Dot15d4FCS  fcf_reserved_1=0 fcf_panidcompress=True fcf_ackreq=True fcf_pending=False fcf_security=False fcf_frametype=Data fcf_srcaddrmode=Short fcf_framever=0 fcf_destaddrmode=Short fcf_reserved_2=0 seqnum=128 |<Dot15d4Data  dest_panid=0x3359 dest_addr=0x18c0 src_addr=0xb7e4 |<ZigbeeNWK  discover_route=0 proto_version=2 frametype=data flags=security+extended_dst+extended_src destination=0x00 source=0xb7e4 radius=10 seqnum=234 ext_dst=ff:f0:00:01:f0:22 ext_src=ff:f0:00:04:15:b1 |<ZigbeeSecurityHeader  reserved1= extended_nonce=1 key_type=network_key nwk_seclevel=None fc=0x730c source=ff:f0:00:04:15:b1 key_seqnum=0 data='[\x9d6\xfc{\x10\t-\xffu,\xe8y\xbb\xcai\x9dR\xc5\xdd\x90\x8b\xd7\x87\xba\xb4/\\\x02:\xd4\xd8F' mic='' |>>>>
   Packet match: OK

and here is a similar packet that fails:

Packet: 15
   Dot15d4FCS / 802.15.4 Data ( 0x00:0xb7e4 -> 0x3359:0x18c0 ) / ZigbeeNWK / ZigbeeSecurityHeader
   <Dot15d4FCS  fcf_reserved_1=0 fcf_panidcompress=True fcf_ackreq=True fcf_pending=False fcf_security=False fcf_frametype=Data fcf_srcaddrmode=Short fcf_framever=0 fcf_destaddrmode=Short fcf_reserved_2=0 seqnum=130 |<Dot15d4Data  dest_panid=0x3359 dest_addr=0x18c0 src_addr=0xb7e4 |<ZigbeeNWK  discover_route=0 proto_version=2 frametype=data flags=security+extended_dst+extended_src destination=0x00 source=0xb7e4 radius=10 seqnum=236 ext_dst=ff:f0:00:01:f0:22 ext_src=ff:f0:00:04:15:b1 |<ZigbeeSecurityHeader  reserved1= extended_nonce=1 key_type=network_key nwk_seclevel=None fc=0x730d source=ff:f0:00:04:15:b1 key_seqnum=0 data=',\xdf\x9c\xd2\x08q\xf7 \xf2\x8a\xb9\xf3\x90\x0b:\xf3C,\x05\x15}\x836b%\xb4\xad\xc1\x02\x8d\xcd\x81Ud\xc6\xaa\x86\xf0y\x03\xb7\r1' |>>>>
   decrypting NWK...
Decrypt Details:
        Key:            26546b723b396a727b5d5271517d392f
        Nonce:          1a5b410000ff0f000d7300002d
        ZigbeeData:     081a0000e4b70aec22021f0000ff0f001a5b410000ff0f002d0d7300001a5b410000ff0f0000
        Decrypted Data: 40c501005cc2c52d3074363437312073612063342e7a722e626220303020303030302030303030
        Encrypted Data: 2cdf9cd20871f720f28ab9f3900b3af3432c05157d83366225b4adc1028dcd815564c6aa86f079
        Mic:            03b70d31
      Raw
      <Raw  load='@\xc5\x01\x00\\\xc2\xc5-0t6471 sa c4.zr.bb 00 0000 0000' |>
      40c501005cc2c52d3074363437312073612063342e7a722e626220303020303030302030303030
   encrypting NWK...
Encrypt Details:
        Key:            26546b723b396a727b5d5271517d392f
        Nonce:          1a5b410000ff0f000d7300002d
        ZigbeeData:     081a0000e4b70aec22021f0000ff0f001a5b410000ff0f002d0d7300001a5b410000ff0f0000
        Decrypted Data: 40c501005cc2c52d3074363437312073612063342e7a722e626220303020303030302030303030
        Encrypted Data: 2cdf9cd20871f720f28ab9f3900b3af3432c05157d83366225b4adc1028dcd815564c6aa86f079
        Mic:            8f2344df
   <Dot15d4FCS  fcf_reserved_1=0 fcf_panidcompress=True fcf_ackreq=True fcf_pending=False fcf_security=False fcf_frametype=Data fcf_srcaddrmode=Short fcf_framever=0 fcf_destaddrmode=Short fcf_reserved_2=0 seqnum=130 |<Dot15d4Data  dest_panid=0x3359 dest_addr=0x18c0 src_addr=0xb7e4 |<ZigbeeNWK  discover_route=0 proto_version=2 frametype=data flags=security+extended_dst+extended_src destination=0x00 source=0xb7e4 radius=10 seqnum=236 ext_dst=ff:f0:00:01:f0:22 ext_src=ff:f0:00:04:15:b1 |<ZigbeeSecurityHeader  reserved1= extended_nonce=1 key_type=network_key nwk_seclevel=None fc=0x730d source=ff:f0:00:04:15:b1 key_seqnum=0 data=',\xdf\x9c\xd2\x08q\xf7 \xf2\x8a\xb9\xf3\x90\x0b:\xf3C,\x05\x15}\x836b%\xb4\xad\xc1\x02\x8d\xcd\x81Ud\xc6\xaa\x86\xf0y\x8f#D\xdf' mic='' |>>>>
   Packet match: FAILED!

Notice the HEX debug for the first NWK layer decrypt (packet 3):

40c501005cc2c52c3074363437302073612063342e7a722e6d6f740d0a

Which is interpreted as "ZigbeeAppDataPayload / ZigbeeClusterLibrary / Raw"

and the second (packet 15):

40c501005cc2c52d3074363437312073612063342e7a722e626220303020303030302030303030

which is interpreted only as "Raw"

**Comment by AdamLaurie** [comment]
My conclusion is this is a false alarm - Wireshark reports these packets as "Bad FCS" and Perytons gets multiple errors so I'm putting this down to Garbage In, Garbage Out!

**Comment by gpotter2** [comment]
The ZigBee fixes have been merged to secdev/scapy

**Comment by rmspeers** [comment]
@AdamLaurie  given the last comments, is this issue closed or does anything else need to be merged from your fork?

File: scripts/zbtestcrypt [hunk 1]
```diff
@@ -13,36 +13,75 @@ from scapy.all import *
 from killerbee import *
 from killerbee.scapy_extensions import *
 
+from rangeparser import RangeParser
+
+# new scapy needs to know if we're using sixlowpan or zigbee
+conf.dot15d4_protocol = "zigbee"
```
[the remark this answers is not in the document]
Comment (AdamLaurie) [reply scripts/zbtestcrypt]: yes, fair enough. will do.

File: scripts/zbtestcrypt [hunk 2]
```diff
@@ -13,36 +13,72 @@ from scapy.all import *
...
+
+    if encrypt:
+        return True, kbencrypt(packet, data, key, verbose= args.verbose)
+    else:
+        return True, kbdecrypt(packet, key, verbose= args.verbose)
+
 if __name__ == '__main__':
     # Command-line arguments
     parser = argparse.ArgumentParser(description="zbtestcrypt: \
         Decrypt, re-encrypt and optionally transmit packets for testing and tool development.")
```
Comment (rmspeers) [inline scripts/zbtestcrypt]: I'm thinking longer term of moving this, along with some test PCAPs, to create a unittest that runs and remove it as a script. Do you have any thoughts as to if this is needed as a script vs a unittest?

File: scripts/zbtestcrypt [hunk 3]
```diff
@@ -51,70 +87,164 @@ if __name__ == '__main__':
...
+                stat, decrypted = True, copy.copy(packet)
+            if stat:
+                print '     ', decrypted.summary()
+                if args.verbose:
+                    print '     ', repr(decrypted)
+                if args.verbose > 1:
+                    print ' HEX:', str(decrypted).encode('hex')
+
+                # If we have an APS layer and it's encrypted, build new packet with ZigbeeNWK so we can decrypt it
+                if decrypted.haslayer(ZigbeeAppDataPayload) and decrypted.haslayer(ZigbeeSecurityHeader) and decrypted[ZigbeeSecurityHeader].data != '':
```
[the remark this answers is not in the document]
Comment (AdamLaurie) [reply scripts/zbtestcrypt]: Yes, I've been thinking the same thing - most of this logic belongs in either scapy_extensions or scapy itself. However, it is an iterative process to get there and to get this all tested and that's why I've put most of the effort into the test script until we've figured out all the 'gotchas' and I was then going to look at migrating. I'm concerned that we now have a partial APS implementation merged into killerbee and most of the fixes sitting here in this PR. I suggest you merge this as it is now as it fixes a lot of issues with the previous merge and then we take it from there?

File: scripts/zbtestcrypt (hunk 2 above)
Comment (AdamLaurie) [reply scripts/zbtestcrypt]: I think we should do both. The script is proving very useful as an analysis tool as well so I'm inclined to keep it (and even move it into the 'tools' section?). I will start crearting sample pcaps with specific features.

**Comment by AdamLaurie** [comment]
Thanks for the merge. I think you can also close https://github.com/riverloopsec/killerbee/pull/66 now.

**Pull request merged**